### DISCUSSION TITLE
Method Stack Heap

### DISCUSSION THREAD

**Discussion by P-C-Space** [NHN-Team-03/ITP#25] [General]
각각의 공간의 개념과 원리를 설명합시다.

**Comment by f1v3-dev** [comment]
> 일단, 자바 프로그램이 실행이 되면 JVM은 OS로 부터 필요한 메모리를 할당받게 됩니다.
> 메모리의 영역은 `Method`, `Stack`, `Heap` 영역으로 나뉘게 됩니다.

아래는 세 영역에 대한 설명입니다.

1. **Method**: 프로그램에서 사용되는 클래스에 대한 정보와 클래스에서 함께 사용되는 정적 변수(static variables)가 저장되는 영역입니다.
JVM은 특정 클래스가 실행된다면 해당 클래스의 파일(*.class)을 읽은 후 해당 클래스에 대한 정보를 가지게 되는 것입니다.

2. **Stack**: Java에서 Method가 호출되면 해당 Method 안에 존재하는 매개변수와 지역변수에 대한 정보가 저장되는 영역입니다.
이와 같이 Method의 호출과 함께 사용되며 이름에서 알 수 있듯이 Stack과 같이 push(Method 호출)가 된 후 pop(Method 호출 완료) 되는 구조입니다.
또한, Heap 영역에 생성된 객체들을 참조하는 주소 값들이 할당됩니다.

3. **Heap**: 모든 인스턴스 객체들이 저장되는 영역으로, 쉽게 생각하면 자바에서 new로 생성한 객체의 정보는 모두 Heap에 저장됩니다. 

---

추가적인 설명이나 지적은 환영입니다 😄

**Reply by Huni0819** [reply]
Stack 영역에 저장된 데이터중에서 한 Method에 의해서 저장되어진 데이터는 해당 Method의 실행이 종료되면 사라집니다. 또한 Stack에 저장 될 수 있는 Data의 크기에는 제한이 있습니다. Heap 영역에 위치한 Object의 주소를 가진 변수가 Stack에 저장될 수 있는데 해당 변수를 reference 변수라고 합니다.

Heap 영역의 데이터는 Stack 영역과 다르게 임의의 위치에 객체가 생성됩니다. 해당 객체를 참조하는 변수가 삭제 되더라도 해당 객체는 삭제되지 않습니다.

**Reply by f1v3-dev** [reply]
### Call Stack의 특징 정리
- 메서드가 호출되면 수행에 **필요한 만큼의 메모리**를 스택에 할당받는다.
- 메서드가 수행을 마치고나면 **사용했던 메모리를 반환**하고 **스택에서 제거**된다.
- 호출스택의 **제일 위**에 있는 메서드가 **현재 실행 중인 메서드**이다.
- 아래에 있는 메서드가 **바로 위의 메서드를 호출한 메서드**이다.
